sdk/helper/ocsp/client.go

```diff
@@ -0,0 +1,1085 @@
+		return res, err
+	}
+
+	ocspHost.Path = ocspHost.Path + "/" + base64.StdEncoding.EncodeToString(reqBody)
+	var res *http.Response
+	request, err := req("GET", ocspHost.String(), nil)
+	if res, err = doRequest(request); err != nil {
+		return nil, nil, nil, err
+	} else {
+		defer res.Body.Close()
+	}
+	if res.StatusCode == http.StatusMethodNotAllowed {
+		request, err = req("POST", origHost.String(), bytes.NewBuffer(reqBody))
+		if res, err = doRequest(request); err != nil {
+			return nil, nil, nil, err
+		} else {
+			defer res.Body.Close()
+		}
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, nil, nil, fmt.Errorf("HTTP code is not OK. %v: %v", res.StatusCode, res.Status)
+	}
+	ocspResBytes, err = io.ReadAll(res.Body)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	ocspRes, err = ocsp.ParseResponse(ocspResBytes, issuer)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return ocspRes, ocspResBytes, &ocspStatus{
+		code: ocspSuccess,
+	}, nil
+}
+
+// GetRevocationStatus checks the certificate revocation status for subject using issuer certificate.
+func (c *Client) GetRevocationStatus(ctx context.Context, subject, issuer *x509.Certificate, conf *VerifyConfig) (*ocspStatus, error) {
+	status, ocspReq, encodedCertID, err := c.ValidateWithCache(subject, issuer)
+	if err != nil {
+		return nil, err
+	}
+	if isValidOCSPStatus(status.code) {
+		return status, nil
+	}
+	if ocspReq == nil || encodedCertID == nil {
+		return status, nil
+	}
+	c.Logger().Debug("cache missed", "server", subject.OCSPServer)
+	if len(subject.OCSPServer) == 0 && len(conf.OcspServersOverride) == 0 {
+		return nil, fmt.Errorf("no OCSP responder URL: subject: %v", subject.Subject)
+	}
+	ocspHosts := subject.OCSPServer
+	if len(conf.OcspServersOverride) > 0 {
+		ocspHosts = conf.OcspServersOverride
+	}
+
+	var wg sync.WaitGroup
+
+	ocspStatuses := make([]*ocspStatus, len(ocspHosts))
+	ocspResponses := make([]*ocsp.Response, len(ocspHosts))
+	errors := make([]error, len(ocspHosts))
+
+	for i, ocspHost := range ocspHosts {
+		u, err := url.Parse(ocspHost)
+		if err != nil {
+			return nil, err
+		}
+
+		hostname := u.Hostname()
+
+		headers := make(map[string]string)
+		headers[httpHeaderContentType] = ocspRequestContentType
+		headers[httpHeaderAccept] = ocspResponseContentType
+		headers[httpHeaderContentLength] = strconv.Itoa(len(ocspReq))
+		headers[httpHeaderHost] = hostname
+		timeout := defaultOCSPResponderTimeout
+
+		ocspClient := retryablehttp.NewClient()
+		ocspClient.HTTPClient.Timeout = timeout
+		ocspClient.HTTPClient.Transport = newInsecureOcspTransport(conf.ExtraCas)
+
+		doRequest := func() error {
+			if conf.QueryAllServers {
+				defer wg.Done()
+			}
+			ocspRes, _, ocspS, err := c.retryOCSP(
+				ctx, ocspClient, retryablehttp.NewRequest, u, headers, ocspReq, issuer)
+			ocspResponses[i] = ocspRes
+			if err != nil {
+				errors[i] = err
+				return err
+			}
+			if ocspS.code != ocspSuccess {
+				ocspStatuses[i] = ocspS
```

Comment:
We have to create a separate closure variable and pass in the value of `i`, right? I think this'd be [the same underlying bug as testing has](https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721#what-happened) -- the value of `i` and `ocspHost` keeps incrementing?

I think this only works in your test since they're all the same CA/URI/... but I'm not really sure. 

(Which is to say, it only happens in the parallel case, not in the serial case).